[feature][processing] A brand new Extract labels algorithm

nirvn opened the pull request:
## Description

This PR adds a brand new Extract labels algorithm to QGIS' processing toolbox. This algorithm extracts label information from a rendered map at a given extent and scale. If a map theme is provided, the rendered map will match the visibility and symbology of that theme. If left blank, the project's layer tree setup will be used. 

Extracted label information include its position (served as point geometries), the associated layer name and feature ID, as well as label rotation, text, alignment, font family, size, weight, etc.

Screenshot of the algorithm dialog:
![image](https://user-images.githubusercontent.com/1728657/148042537-c4248d54-b6cf-4e9c-8ccc-47debc9190e9.png)

Screenshot showcasing an output of the algorithm:
![image](https://user-images.githubusercontent.com/1728657/147735563-5574791e-430f-4c97-8db0-417270b12f8b.png)

The algorithm also generates a default symbol-less & label style for the output layer which allows for those saved layers to be drag and dropped into projects and have the label drawn exactly as they were exported.

https://user-images.githubusercontent.com/1728657/147822919-ddc2b698-7565-4d58-921f-b58e68aeec46.mp4

_Funded by [TEKSI](https://www.teksi.ch/)_

nirvn committed:
[feature][processing] A brand new Extract labels algorithm

nirvn committed:
[api] Export the core QgsTextLabelFeature class

m-kuhn commented on src/analysis/processing/qgsalgorithmextractlabels.cpp:
```diff
@@ -0,0 +1,364 @@
+{
+  return QObject::tr( "This algorithm extracts label information from a rendered map at a given extent and scale.\n\n"
+                      "If a map theme is provided, the rendered map will match the visibility and symbology of that theme. If left blank, the project's layer tree setup will be used.\n\n"
+                      "Extracted label information include position, font family, size, weight, rotation, and more." );
```
worth mentionning: source layer, feature id, label text

m-kuhn commented on src/analysis/processing/qgsalgorithmextractlabels.cpp:
```diff
@@ -0,0 +1,364 @@
+
+      QgsTextLabelFeature *labelFeature = dynamic_cast<QgsTextLabelFeature *>( label->getFeaturePart()->feature() );
+      if ( !labelFeature )
+        return;
```
Worth a message?

nirvn replied:
Actually, after looking at other parts of our code, this condition can be removed safely.

nyalldawson commented on src/analysis/processing/qgsalgorithmextractlabels.cpp:
```diff
@@ -0,0 +1,372 @@
+  qgsalgorithmextractlabels.cpp - QgsExtractLabelsAlgorithm
+
+ ---------------------
+ Original implementation in Python:
+
+ begin                : 2016-10-05
+ copyright            : (C) 2016 by OPENGIS.ch
+ email                : matthias@opengis.ch
+
+ C++ port:
+
+ begin                : 20.11.2019
+ copyright            : (C) 2019 by Alessandro Pasotti
```
these lines look wrong

nyalldawson commented on src/analysis/processing/qgsalgorithmextractlabels.cpp:
```diff
@@ -0,0 +1,372 @@
+
+QStringList QgsExtractLabelsAlgorithm::tags() const
+{
+  return QObject::tr( "map themes" ).split( ',' );
```
was this a placeholder?

nyalldawson commented on src/analysis/processing/qgsalgorithmextractlabels.cpp:
```diff
@@ -0,0 +1,372 @@
+
+QString QgsExtractLabelsAlgorithm::group() const
+{
+  return QObject::tr( "Vector creation" );
```
I'd suggest the "Cartography" group is more appropriate

nyalldawson commented on src/analysis/processing/qgsalgorithmextractlabels.cpp:
```diff
@@ -0,0 +1,372 @@
+                  QgsProcessingParameterNumber::Double,
+                  QVariant( 96.0 ) ) );
+
+  addParameter( new QgsProcessingParameterMapTheme(
```
This parameter could do with the extended help, eg explaining what it means when its unset

nyalldawson commented on src/analysis/processing/qgsalgorithmextractlabels.cpp:
```diff
@@ -0,0 +1,372 @@
+
+QString QgsExtractLabelsAlgorithm::shortDescription() const
+{
+  return QObject::tr( "Extract label information (position, font details, etc.) from a rendered map at a given extent and scale." );
```
How about something more concise here, like "Converts map labels to a point layer" ?

nyalldawson commented on src/analysis/processing/qgsalgorithmextractlabels.cpp:
```diff
@@ -0,0 +1,372 @@
+{
+  return QObject::tr( "This algorithm extracts label information from a rendered map at a given extent and scale.\n\n"
+                      "If a map theme is provided, the rendered map will match the visibility and symbology of that theme. If left blank, the project's layer tree setup will be used.\n\n"
+                      "Extracted label information include its position (served as point geometries), the associated layer name and feature ID, as well as label text, rotation, alignment, and font details." );
```
```suggestion
                      "Extracted label information include their position (served as point geometries), the associated layer name and feature ID, as well as label text, rotation, alignment, and font details." );
```

nyalldawson commented on src/analysis/processing/qgsalgorithmextractlabels.cpp:
```diff
@@ -0,0 +1,372 @@
+        return;
+      }
+
+      QgsTextLabelFeature *labelFeature = dynamic_cast<QgsTextLabelFeature *>( label->getFeaturePart()->feature() );
```
```suggestion
      QgsTextLabelFeature *labelFeature = qgis::down_cast<QgsTextLabelFeature *>( label->getFeaturePart()->feature() );
```

nyalldawson replied:
(but wasn't the check needed to correctly handle diagrams? I'd suggest leaving the dynamic cast and putting a nullptr check after this)

nirvn replied:
@nyalldawson , other parts of our code, including QgsVectorLayerLabelProvider::drawLabel (https://github.com/qgis/QGIS/blob/master/src/core/labeling/qgsvectorlayerlabelprovider.cpp#L354) don't have that check, the QgsLabelingEngine::drawLabels only throws labels there, diagrams take a different path.

I had the check there before, but then thought if QgsVectorLayerLabelProvider doesn't need it, it's probably superfluous :)

nyalldawson replied:
I think we should resurrect the dynamic_cast and check here -- it's a trivial check, and worth it for even the remote possibility it avoids a crash in future

nyalldawson commented on src/analysis/processing/qgsalgorithmextractlabels.cpp:
```diff
@@ -0,0 +1,372 @@
+      if ( settings.placement == QgsPalLayerSettings::Curved ||
+           settings.placement == QgsPalLayerSettings::PerimeterCurved )
+      {
+        mFeedback->pushWarning( QObject::tr( "Curved placement not supported, skipping label for FID %1 from layer %2" ).arg( fid ).arg( layerId ) );
```
better to use layer name here rather than id. I'd also suggest that we could drop the feature id from the warning and add logic to ensure this warning is only ever raised once per layer (given that the curved setting is per layer, warning per feature isn't useful, and will flood the log till the point of unusability)

nyalldawson commented on src/analysis/processing/qgsalgorithmextractlabels.cpp:
```diff
@@ -0,0 +1,372 @@
+      QgsPalLayerSettings labelSettings( settings );
+      const QMap< QgsPalLayerSettings::Property, QVariant > &dataDefinedValues = labelFeature->dataDefinedValues();
+
+      if ( dataDefinedValues.contains( QgsPalLayerSettings::MultiLineWrapChar ) )
+      {
+        labelSettings.wrapChar = dataDefinedValues.value( QgsPalLayerSettings::MultiLineWrapChar ).toString();
+      }
+      if ( dataDefinedValues.contains( QgsPalLayerSettings::AutoWrapLength ) )
+      {
+        labelSettings.autoWrapLength = dataDefinedValues.value( QgsPalLayerSettings::AutoWrapLength ).toInt();
+      }
+      const QString labelText = QgsPalLabeling::splitToLines( labelFeature->text( -1 ),
+                                labelSettings.wrapChar,
+                                labelSettings.autoWrapLength,
+                                labelSettings.useMaxLineLengthForAutoWrap ).join( "\n" );
```
Could this logic be moved to QgsTextLabelFeature?

nirvn replied:
@nyalldawson , I'm unsure, see https://github.com/qgis/QGIS/blob/master/src/core/labeling/qgsvectorlayerlabelprovider.cpp#L659 . If we change that, we'd have to apply changes in other parts of the code too. 

One thing this work revealed is how much improvements we could do in the overall labeling code workflow, but maybe that's better done in a separate PR.

nyalldawson commented on src/analysis/processing/qgsalgorithmextractlabels.cpp:
```diff
@@ -0,0 +1,372 @@
+      const QFont font = labelFeature->definedFont();
+      const QString fontFamily = font.family();
+      const QString fontStyle = font.styleName();
+      const double fontSize = static_cast<double>( font.pixelSize() ) * 72 / context.painter()->device()->logicalDpiX();
```
where's the 72 from?

nirvn replied:
@nyalldawson , 72 was what I've been using to do pixel to point conversion. I found that the labelFeature->definedFont() didn't have a proper pointSize() value (returns -1), and I don't think we should store pixel size value as it doesn't take scale in consideration.

nyalldawson replied:
Why 72 though? Shouldn't this be at least 96? (We really should store the text format in QgsTextLabelFeature instead!)

nyalldawson commented on src/analysis/processing/qgsalgorithmextractlabels.cpp:
```diff
@@ -0,0 +1,372 @@
+
+      double x = label->getX();
+      double y = label->getY();
+      if ( dataDefinedValues.contains( QgsPalLayerSettings::PositionX ) )
+      {
+        x = dataDefinedValues.value( QgsPalLayerSettings::PositionX ).toDouble();
+      }
+      if ( dataDefinedValues.contains( QgsPalLayerSettings::PositionY ) )
+      {
+        y = dataDefinedValues.value( QgsPalLayerSettings::PositionY ).toDouble();
+      }
```
this shouldn't be required -- pal has already taken it into account

nirvn replied:
@nyalldawson , in my tests, I saw that label->getX() and  label->getY() would return the non manually set position, hence why this was needed.

nyalldawson replied:
I'm almost 100% it's not required -- otherwise we'd see similar logic in place for rendering labels, and things like QgsLabelSearchTree::insertLabel would be broken for data-defined placements

nyalldawson commented on src/analysis/processing/qgsalgorithmextractlabels.cpp:
```diff
@@ -0,0 +1,372 @@
+      features << feature;
+    };
+
+    QMap<QString, QString> *mMapLayerNames = nullptr;
```
```suggestion
    QMap<QString, QString> mMapLayerNames;
```

There's no reason to use a pointer here

nyalldawson commented on src/analysis/processing/qgsalgorithmextractlabels.cpp:
```diff
@@ -0,0 +1,372 @@
+QVariantMap QgsExtractLabelsAlgorithm::processAlgorithm( const QVariantMap &parameters, QgsProcessingContext &context, QgsProcessingFeedback *feedback )
+{
+  // MAP_THEME has been taken care of in prepareAlgorithm
+  QgsCoordinateReferenceSystem crs = parameterAsExtentCrs( parameters, QStringLiteral( "EXTENT" ), context );
+  if ( !crs.isValid() )
+    crs = context.project()->crs();
```
this should be in prepareAlgorithm (it's potentially not thread safe to access the project crs here, and it's trivial to move to the prepare step)

nyalldawson commented on src/analysis/processing/qgsalgorithmextractlabels.cpp:
```diff
@@ -0,0 +1,372 @@
+  fields.append( QgsField( QStringLiteral( "featureid" ), QVariant::LongLong, QString(), 20 ) );
+  fields.append( QgsField( QStringLiteral( "label_text" ), QVariant::String, QString(), 0, 0 ) );
+  fields.append( QgsField( QStringLiteral( "label_alignment" ), QVariant::String, QString(), 0, 0 ) );
+  fields.append( QgsField( QStringLiteral( "label_rotation" ), QVariant::Double, QString(), 20, 10 ) );
```
20/10 seems overkill for precision here (same with the font size, letter spacing, etc)

nirvn replied:
OK, I've updated this to match the precision to what we show in our UI.

nyalldawson commented on src/analysis/processing/qgsalgorithmextractlabels.cpp:
```diff
@@ -0,0 +1,372 @@
+  mapSettings.setOutputSize( imageSize.toSize() );
+  mapSettings.setOutputDpi( dpi );
+  mapSettings.setFlag( Qgis::MapSettingsFlag::DrawLabeling, true );
+  mapSettings.setFlag( Qgis::MapSettingsFlag::SkipSymbolRendering, true );
```
I think you should also set the UseRenderingOptimization flag here -- that will speed up the algorithm a LOT

nirvn replied:
Great piece of wisdom :)

nyalldawson commented on src/analysis/processing/qgsalgorithmextractlabels.cpp:
```diff
@@ -0,0 +1,372 @@
+  feedback->pushInfo( QObject::tr( "Extracting labels" ) );
+  renderJob.start();
+  renderJob.waitForFinished();
+  qDeleteAll( mMapLayers );
```
```suggestion
  qDeleteAll( mMapLayers );
  mMapLayers.clear();
```

nyalldawson commented on src/analysis/processing/qgsalgorithmextractlabels.cpp:
```diff
@@ -0,0 +1,372 @@
+  qDeleteAll( mMapLayers );
+
+  feedback->pushInfo( QObject::tr( "Writing %n label(s) to output layer", "", labelSink.features.count() ) );
+  for ( QgsFeature feature : std::as_const( labelSink.features ) )
```
```suggestion
  for ( const QgsFeature &feature : std::as_const( labelSink.features ) )
```

nyalldawson commented on src/analysis/processing/qgsalgorithmextractlabels.cpp:
```diff
@@ -0,0 +1,372 @@
+  if ( !mapTheme.isEmpty() && context.project()->mapThemeCollection()->hasMapTheme( mapTheme ) )
+  {
+    const QList<QgsMapLayer *> constLayers = context.project()->mapThemeCollection()->mapThemeVisibleLayers( mapTheme );
+    for ( const auto *l : constLayers )
```
```suggestion
    for ( const QgsMapLayer *l : constLayers )
```

nyalldawson commented on src/analysis/processing/qgsalgorithmextractlabels.cpp:
```diff
@@ -0,0 +1,372 @@
+    mMapThemeStyleOverrides = context.project()->mapThemeCollection( )->mapThemeStyleOverrides( mapTheme );
+  }
+
+  if ( mMapLayers.size() == 0 )
```
```suggestion
  if ( mMapLayers.empty() )
```

nyalldawson commented on src/analysis/processing/qgsalgorithmextractlabels.cpp:
```diff
@@ -0,0 +1,372 @@
+    mMapLayerNames.insert( l->id(), l->name() );
+  }
+
+  return mMapLayers.size() > 0;
```
```suggestion
  return true;
```

nyalldawson commented on src/analysis/processing/qgsalgorithmextractlabels.h:
```diff
@@ -0,0 +1,68 @@
+
+ ---------------------
+
+ Original implementation in Python:
```
header needs updating too

nyalldawson commented:
Nice work!

I have a couple of general (non code) comments too:

- The algorithm should automatically set the style for the generated layer appropriately, so that the layer is already labelled when its output from the algorithm. This style should also be set as the default style for the layer too, so that loading the layer into a different project results in the labels appearing immediately as expected. A null symbol renderer should be used here so that only labels are shown. Also the symbology reference scale for the layer style should be set to the scale used to generate the labels.
- The algorithm should expose options for whether all labels (including conflicting ones) are output (and this should probably be the default setting). Right now you'll only get non-conflicting labels and no way of knowing if labels are missing.
- Similarly, I think advanced settings should be added to expose the label settings "number of candidates" options. These can substantially speed up labelling, and if someone wants to pre-generate labelling e.g. for a whole country then lowering the candidate counts will become a necessary tweak.
- Needs support added for cancelling the job early (by hooking the algorithm's feedback cancellation into the map job's feedback, and also early exit paths for the label to feature conversion code)

Exciting feature, nice work!

nyalldawson commented on src/analysis/processing/qgsalgorithmextractlabels.cpp:
```diff
@@ -0,0 +1,372 @@
+  fields.append( QgsField( QStringLiteral( "label_text" ), QVariant::String, QString(), 0, 0 ) );
+  fields.append( QgsField( QStringLiteral( "label_alignment" ), QVariant::String, QString(), 0, 0 ) );
+  fields.append( QgsField( QStringLiteral( "label_rotation" ), QVariant::Double, QString(), 20, 10 ) );
+  fields.append( QgsField( QStringLiteral( "font_family" ), QVariant::String, QString(), 0, 0 ) );
```
Can we use the equivalent name from the matching data defined property here for the field names? This would then make it possible to easily create fields from ALL labeling settings in future without breaking models and other workflows.

nirvn committed:
[api] add unplaced label handling to the QgsLabelSink class

nirvn commented:
@nyalldawson , thanks for the extensive review, that made the algorithm substantially better.

I think I've addressed all your review points (left a couple of comments). I've also implemented the unplaced label option as I think it's a really useful idea ( :partying_face: ) that the alg's basic implementation should cover.

As for the style generation, great idea, but I'd prefer leaving that to a follow up PR as it goes beyond the scope of the work this PR covers. OK with you?

nirvn commented:
Styling actually implemented, @nyalldawson offered good rationals for it.

DelazJ commented:
Great addition, @nirvn 
I don't know if it has been addressed in reviews but what does `map scale = 0` in the screenshot mean? Maybe, fill it with current scale by default?
For changelog, could you also update the alg screenshot, please (it seems you added few things, in the meantime)?

nirvn commented:
@DelazJ , I'll update the screenshot in a minute.

The scale widget puts 0 as default value (same as what you'd see when setting the label scale dependency visibility:
![image](https://user-images.githubusercontent.com/1728657/147810069-d7ac54cb-9dff-451c-850e-b1ecc4123c5c.png)

I've now made it so the algorithm will throw an error if someone inputs 0 as scale.

m-kuhn commented on src/analysis/processing/qgsalgorithmextractlabels.cpp:
```diff
@@ -0,0 +1,461 @@
+        case QgsPalLayerSettings::MultiFollowPlacement:
+          switch ( label->getQuadrant() )
+          {
+            case pal::LabelPosition::QuadrantAboveLeft:
```
Would it be possible to get the vertical alignment as well?

nirvn replied:
The algorithm hard-codes the anchor / text block alignment to be bottom left. Is that something we'd want to handle differently?

m-kuhn commented on src/analysis/processing/qgsalgorithmextractlabels.cpp:
```diff
@@ -0,0 +1,461 @@
+{
+  return QObject::tr( "This algorithm extracts label information from a rendered map at a given extent and scale.\n\n"
+                      "If a map theme is provided, the rendered map will match the visibility and symbology of that theme. If left blank, the project's layer tree setup will be used.\n\n"
+                      "Extracted label information include their position (served as point geometries), the associated layer name and feature ID, as well as label text, rotation, alignment, and font details." );
```
```suggestion
                      "Extracted label information include\n"
                      " - position (served as point geometries)\n"
                      " - the associated layer name and feature ID\n"
                      " - label text\n"
                      " - rotation in degree, clockwise\n"
                      " - alignment\n"
                      " - font details" );
```

Can you check the rotation (cw/ccw) and adjust alignment according to my other comment where required?

nirvn replied:
@m-kuhn , rotation details added (it was actually missing from the default style attached to the output layer, good catch), and description tweaked. I haven't used a bullet list, that doesn't play well with the formatting we apply to short help strings.

m-kuhn replied:
We should support md here :-)

nirvn replied:
@m-kuhn , yeah, it'd be nice (would likely output OK as ASCII for qgis_process too)

nyalldawson commented on src/analysis/processing/qgsalgorithmextractlabels.cpp:
```diff
@@ -0,0 +1,464 @@
+        QStringLiteral( "MAP_THEME" ),
+        QObject::tr( "Map theme" ),
+        QVariant(), true );
+  mapThemeParameter->setHelp( QObject::tr( "This parameter is optional. When left unset, the algorithm will fallback to match the currently opened project's layer tree visibility to extract labels from." ) );
```
```suggestion
  mapThemeParameter->setHelp( QObject::tr( "This parameter is optional. When left unset, the algorithm will fallback to extracting labels from all currently visible layers in the project." ) );
```

"layer tree" is an internal name, we never expose this to users

nyalldawson commented on src/analysis/processing/qgsalgorithmextractlabels.cpp:
```diff
@@ -0,0 +1,464 @@
+QString QgsExtractLabelsAlgorithm::shortHelpString() const
+{
+  return QObject::tr( "This algorithm extracts label information from a rendered map at a given extent and scale.\n\n"
+                      "If a map theme is provided, the rendered map will match the visibility and symbology of that theme. If left blank, the project's layer tree setup will be used.\n\n"
```
```suggestion
                      "If a map theme is provided, the rendered map will match the visibility and symbology of that theme. If left blank, all visible layers from the project will be used.\n\n"
```

nyalldawson commented on src/analysis/processing/qgsalgorithmextractlabels.cpp:
```diff
@@ -0,0 +1,464 @@
+
+  QEventLoop loop;
+  QObject::connect( feedback, &QgsFeedback::canceled, &renderJob, &QgsMapRendererCustomPainterJob::cancel );
+  QObject::connect( &renderJob, &QgsMapRendererJob::renderingLayersFinished, feedback, [feedback]() { feedback->pushInfo( QObject::tr( "Calculating label placement" ) ); } );
```
```suggestion
  QObject::connect( &renderJob, &QgsMapRendererJob::renderingLayersFinished, feedback, [feedback]() { feedback->pushInfo( QObject::tr( "Calculating label placement" ) ); } );
  QObject::connect( &renderJob, &QgsMapRendererJob::layerRenderingStarted, feedback, [this, feedback]( const QString& layerId ) { feedback->pushInfo( QObject::tr( "Collecting labels for %1" ).arg( mMapLayerNames.value( layerId ); } );
```

nirvn commented:
@nyalldawson , all comments addressed, minus the 72 value, which is needed for the algorithm to produce point values that match what's drawn on the canvas. Let's give that a try on people's systems.
